### DISCUSSION TITLE
Which PostgreSQL version to use ?

### DISCUSSION THREAD

**Discussion by Rollbacke** [dani-garcia/vaultwarden#3960] [Q&A]
Hi everyone,
Pretty simple question to answer actually.
Today, my db for vaultwwarden is on a "postgres:15" docker image.
Version 16 exists today. Is it supported/tested by anyone ?
Thanks :)

**Comment by BlackDex** [comment]
Vaultwarden is build using the v15 libraries.
So, if the v16 server supports v15 clients, then it should be all fine, if there is an issue with that, then i suggest to use a v15 server.

I have not yet tested it my self, but maybe there is something mentioned on the PostgreSQL release page in regards to backwards compatibility with older clients?

**Comment by Rollbacke** [comment]
Hi @BlackDex,
According to https://www.postgresql.org/about/news/postgresql-16-released-2715/ and https://www.postgresql.org/support/versioning/ , I think there is no problem.
At least, nothing is cited in this sense.

I will try on my install.

**Comment by Rollbacke** [comment]
It seems to work correctly.
Accessible from the web page and from a bitwarden on Linux/Windows.
Test of adding login/password OK.

**Comment by kevdogg** [comment]
I'm currently running vaultwarden with a bitnami postgresql container which is version at 11.14.0.  Just doing some routine maintainence on my container stack and I see this version is really old.  I'm not postgresql expert, but do you either have a link or process by which you would upgrade the database?  I've never done this before however I'm guessing its an export of the data, start the new database container version, create tables and then import the data.  Just wanted to verify steps as this is new to me.

**Reply by BlackDex** [reply]
Best to ask these questions at a PostgreSQL forum or something.
The probably are some how to upgrade PostgreSQL version blogs out there too.

**Comment by ccben87** [comment]
@Rollbacke & anyone else:

If you're on the Postgres 16 then make sure you're running at least 16.2. Possibly not a good idea to be running 16.x at all given the issue we ran into. We have found out the hard way that using Vaultwarden with 16.0 (and I presume 16.1) would cause us to receive cached failed errors for certain users or for certain login attempts. Initially we did not have the issue but after some time of using Vaultwarden we had the issue happen at which point we had to restore from backup. We later had the issue occur again and this time when investigating we found a bug was reported for the issue we were having and fixed in 14.2: https://git.postgresql.org/gitweb/?p=postgresql.git;a=commitdiff;h=74f770ef202fd3f1f1d1f0889f04c16440d23bd0

If you're on 16.0 or 16.1, I would highly recommend you update to 16.2 or there's a good chance you will hit the bug eventually like we did.

**Reply by BlackDex** [reply]
As a side-note, Vaultwarden uses v15.x as the client library of PostgreSQL.
At least for the containers we build.

**Reply by Rollbacke** [reply]
Hi @ccben87 ,
In October, my database was probably running on PostgreSQL < 16.2 (because it was probably only published on Docker a month ago, https://github.com/docker-library/postgres/commits/master/16/bookworm). I haven't experienced any problems, however, I no longer have the logs to prove it...
Anyway, I'm using PostgreSQL 16.2 (I use the tag ":latest", so "PG_VERSION=16.2-1.pgdg120+2") today.
I confirm, everything is fine on my side.

**Comment by logicoa** [comment]
hey @Rollbacke still no problem with using postgres:latest config?

**Reply by kevdogg** [reply]
@logicoa I'd be careful about using latest. Postgres isn't good at handling major version upgrades sometimes..ie 14 to 15 or 16 to 17.  In those cases it's best to go through the formal upgrade process..not saying sometimes things might not work but it's not recommended since sometime things will break. Minor upgrades are fine.  It's probably best to pin your postgres version to prevent a major version upgrade by accident.

@BlackDex . On the vaultwarden wiki is the client postgres version libraries listed? I'm currently using postgres 17.x without issues however I only have a few clients so my use case probably isn't adequate for testing.

**Reply by Rollbacke** [reply]
Hey @logicoa,
As @kevdogg said, I don't (or no longer) use the latest tag, because major PostgreSQL updates aren't done automatically (backup/restore needed), and sometimes you're surprised to realize it when you need it -_-'
I fixed the version : currently running PSQL 16 without any problems. I admit I haven't tested it in PostgreSQL 17.
(Which version is officially supported on Vaultwarden today?)

**Reply by BlackDex** [reply]
Our current Vaultwarden images use v15 and to be specific v15.3 for Debian based images and depending on when the images are built most likely the latest minor version of v15 for the Alpine images. I built just today the latest v15.13.

When Debian Trixie is being released it will use v17, and then we probably also switch to use v17 for Alpine since i already built v17, v16 and v15 in the Alpine build images.

Also, as far as i know, clients/servers are compatible with older/newer versions, except that some features might not work of course. But those are probably to specific for PostgreSQL for us anyway.

**Reply by Rollbacke** [reply]
@BlackDex Thanks for the clarification ! 😊

**Reply by logicoa** [reply]
Thank you guys for clarification :)
